<a href="https://colab.research.google.com/github/Kuriajm/DS_IP_Wk3/blob/main/Moringa_Data_Science_Prep_W3_Independent_Project_2021_09_Joseph_Mutura_Pandas_ipynb.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Week 3 IP-Mutura Kuria

How does MTN Cote d'Ivoire go about the upgrade of its infrastructure strategy within the given cities?

# 1.0 Importing Pandas and Numpy

In [1]:
# OUR CODE GOES HERE
import pandas as pd
# as well as the Numpy library
import numpy as np

#1.1 Loading our Datasets

In [3]:
#Loading Dataset for day1
url = '/content/CDR_20120507.csv'
dfd1 = pd.read_csv(url) 
dfd1.head(10)

PRODUTC  VALUE               DATETIME  ... COUNTRY_B     CELL_ID     SITE_ID
0   Voice      0  2012-05-06 23:04:37.0  ...     19e2e  /TJNe+Mmtu  /TJNe+Mmtu
1     sms      0  2012-05-06 23:05:25.0  ...     19e2e  /+cKZKIp41  /+cKZKIp41
2     sms     25  2012-05-06 23:05:41.0  ...     19e2e  /u0FSD+ahi  /u0FSD+ahi
3   Voice     56  2012-05-06 23:05:42.0  ...     19e2e  +cNeJzsTp3  +cNeJzsTp3
4   Voice     14  2012-05-06 23:06:29.0  ...     19e2e  051KhYDCpv  051KhYDCpv
5    data      0  2012-05-06 23:07:26.0  ...     OTHER  /lrUXVGVcN  /lrUXVGVcN
6   Voice      0  2012-05-06 23:08:02.0  ...     19e2e  +38u2u/rfx  +38u2u/rfx
7   Voice      0  2012-05-06 23:08:03.0  ...     19e2e  09UjmmY1Ds  09UjmmY1Ds
8   Voice      0  2012-05-06 23:08:05.0  ...     19e2e  0DeimCpZOc  0DeimCpZOc
9     sms      0  2012-05-06 23:08:43.0  ...     19e2e  01MDiY5F89  01MDiY5F89

[10 rows x 10 columns]

In [4]:
#Loading Dataset for day 2
url = '/content/CDR_20120508.csv'
dfd2 = pd.read_csv(url) 
dfd2.head() 

PRODUCT  VALUE              DATE_TIME  ... COUNTRY_B     CELL_ID     SITE_ID
0     sms      0  2012-05-07 23:02:06.0  ...     19e2e  +854AcBQT2  +854AcBQT2
1   Voice      0  2012-05-07 23:03:44.0  ...     19e2e  +854AcBQT2  +854AcBQT2
2     sms      0  2012-05-07 23:04:06.0  ...     19e2e  +laSrk7g6q  +laSrk7g6q
3   Voice     37  2012-05-07 23:04:48.0  ...     19e2e  /xvRrCVKoQ  /xvRrCVKoQ
4     sms     25  2012-05-07 23:04:58.0  ...     19e2e  +bO+qkSonO  +bO+qkSonO

[5 rows x 10 columns]

In [5]:
#Loading Dataset for day 3
url = '/content/CDR_20120509.csv'
dfd3 = pd.read_csv(url) 
dfd3.head()

PRODUCT  VALUE              DATE_TIME  ... COUNTRY_B      CELLID     SIET_ID
0   Voice     61  2012-05-08 23:01:28.0  ...     19e2e  +38u2u/rfx  +38u2u/rfx
1     sms      0  2012-05-08 23:01:31.0  ...     19e2e  +Fs4mTvdKx  +Fs4mTvdKx
2   Voice     10  2012-05-08 23:02:59.0  ...     19e2e  +ZaaZozy/+  +ZaaZozy/+
3     sms      0  2012-05-08 23:03:04.0  ...     19e2e  +cNeJzsTp3  +cNeJzsTp3
4   Voice     22  2012-05-08 23:03:16.0  ...     19e2e  /eeMklgyJA  /eeMklgyJA

[5 rows x 10 columns]

In [16]:
#Loading Dataset for cells Geo
url = '/content/cells_geo.xlsx'
dfcellsgeo = pd.read_excel(url) 
dfcellsgeo

VILLES      STATUS LOCALISATION  ...          AREA     CELL_ID   SITE_CODE
0       ADJAME  In Service      ABIDJAN  ...           NaN  fbb2e75af3  0bb492ef17
1       ADJAME  In Service      ABIDJAN  ...           NaN  1502501d78  0bb492ef17
2       ADJAME  In Service      ABIDJAN  ...           NaN  0c2e182a0c  0bb492ef17
3      ASSINIE  In Service    INTERIEUR  ...           NaN  69bdfc69fd  8a1e6ce31f
4      ASSINIE  In Service    INTERIEUR  ...           NaN  7a1029d2ab  8a1e6ce31f
...        ...         ...          ...  ...           ...         ...         ...
3969  ZUENOULA  In Service    INTERIEUR  ...  CENTRE NORTH  1cfc2d7ab1  f91ed8d2e9
3970  ZUENOULA  In Service    INTERIEUR  ...  CENTRE NORTH  3ae1708153  f91ed8d2e9
3971  ZUENOULA  In Service    INTERIEUR  ...  CENTRE NORTH  cb71f8a29c  61bee9c70c
3972  ZUENOULA  In Service    INTERIEUR  ...  CENTRE NORTH  d4a6a0b645  61bee9c70c
3973  ZUENOULA  In Service    INTERIEUR  ...  CENTRE NORTH  f9f3e42589  61bee9c70c

[3974 rows x 11 columns]

# 1.2 Merging the dataframes

Merge the dataframes so that we have daily data with their locations together

In [12]:
#Before merging the dataframes lets do some cleaning to remove null rows in all the dataframes

dfd1v1 = dfd1.dropna()
dfd1v1

PRODUTC  VALUE               DATETIME  ... COUNTRY_B     CELL_ID     SITE_ID
0      Voice      0  2012-05-06 23:04:37.0  ...     19e2e  /TJNe+Mmtu  /TJNe+Mmtu
1        sms      0  2012-05-06 23:05:25.0  ...     19e2e  /+cKZKIp41  /+cKZKIp41
2        sms     25  2012-05-06 23:05:41.0  ...     19e2e  /u0FSD+ahi  /u0FSD+ahi
3      Voice     56  2012-05-06 23:05:42.0  ...     19e2e  +cNeJzsTp3  +cNeJzsTp3
4      Voice     14  2012-05-06 23:06:29.0  ...     19e2e  051KhYDCpv  051KhYDCpv
...      ...    ...                    ...  ...       ...         ...         ...
4996     sms    100  2012-05-07 00:01:54.0  ...     OTHER  0c2371d9bc  1e751d08a4
4997   Voice      1  2012-05-07 00:01:54.0  ...     19e2e  5fb791cf56  65efe2bc62
4998   Voice      0  2012-05-07 00:01:54.0  ...     19e2e  7773627c12  9ef9616350
4999   Voice      0  2012-05-07 00:01:54.0  ...     19e2e  3f4af49d6c  969b8d0e39
5000   Voice      7  2012-05-07 00:01:54.0  ...     19e2e  6f4d9b77a5  ebfc58af70

[4285 rows x 10 columns]

In [13]:
dfd2v1 = dfd2.dropna()
dfd2v1

PRODUCT  VALUE              DATE_TIME  ... COUNTRY_B     CELL_ID     SITE_ID
0        sms      0  2012-05-07 23:02:06.0  ...     19e2e  +854AcBQT2  +854AcBQT2
1      Voice      0  2012-05-07 23:03:44.0  ...     19e2e  +854AcBQT2  +854AcBQT2
2        sms      0  2012-05-07 23:04:06.0  ...     19e2e  +laSrk7g6q  +laSrk7g6q
3      Voice     37  2012-05-07 23:04:48.0  ...     19e2e  /xvRrCVKoQ  /xvRrCVKoQ
4        sms     25  2012-05-07 23:04:58.0  ...     19e2e  +bO+qkSonO  +bO+qkSonO
...      ...    ...                    ...  ...       ...         ...         ...
4995     sms     25  2012-05-08 00:01:53.0  ...     19e2e  0e8763bdaa  eefd93f0b4
4996     sms      0  2012-05-08 00:01:53.0  ...     19e2e  896459aace  b7d00e59e4
4997     sms     25  2012-05-08 00:01:53.0  ...     19e2e  7e90b35166  211015b14d
4998     sms      0  2012-05-08 00:01:53.0  ...     19e2e  fe70e431df  c7ea989264
5000   Voice      0  2012-05-08 00:01:53.0  ...     19e2e  8630058aca  b47a998a3e

[4325 rows x 10 columns]

In [14]:
dfd3v1 = dfd3.dropna()
dfd3v1

PRODUCT  VALUE              DATE_TIME  ... COUNTRY_B      CELLID     SIET_ID
0      Voice     61  2012-05-08 23:01:28.0  ...     19e2e  +38u2u/rfx  +38u2u/rfx
1        sms      0  2012-05-08 23:01:31.0  ...     19e2e  +Fs4mTvdKx  +Fs4mTvdKx
2      Voice     10  2012-05-08 23:02:59.0  ...     19e2e  +ZaaZozy/+  +ZaaZozy/+
3        sms      0  2012-05-08 23:03:04.0  ...     19e2e  +cNeJzsTp3  +cNeJzsTp3
4      Voice     22  2012-05-08 23:03:16.0  ...     19e2e  /eeMklgyJA  /eeMklgyJA
...      ...    ...                    ...  ...       ...         ...         ...
4996   Voice      0  2012-05-09 00:01:43.0  ...     19e2e  2aa04dcab7  f7d63a29e5
4997   Voice     62  2012-05-09 00:01:43.0  ...     19e2e  7f1f2f889f  b4ddb9e93c
4998   Voice     32  2012-05-09 00:01:43.0  ...     19e2e  a4731e780e  d68260ef50
4999   Voice     30  2012-05-09 00:01:43.0  ...     19e2e  98ed726da2  b4130b54ab
5000   Voice     60  2012-05-09 00:01:43.0  ...     19e2e  1c11a2fcbb  6d70dae094

[4396 rows x 10 columns]

In [15]:
#On geo cell dataset

dfcellsgeov1 = dfcellsgeo.dropna()
dfcellsgeov1

VILLES      STATUS LOCALISATION  ...           AREA     CELL_ID   SITE_CODE
23       ABOBO  In Service      ABIDJAN  ...  ABIDJAN NORTH  f740406877  0f34c36ac5
24       ABOBO  In Service      ABIDJAN  ...  ABIDJAN NORTH  959e54c8f7  0f34c36ac5
25       ABOBO  In Service      ABIDJAN  ...  ABIDJAN NORTH  56b7e7000c  0f34c36ac5
26       ABOBO  In Service      ABIDJAN  ...  ABIDJAN NORTH  bc51c17a6f  0f34c36ac5
27       ABOBO  In Service      ABIDJAN  ...  ABIDJAN NORTH  e460975c9f  0f34c36ac5
...        ...         ...          ...  ...            ...         ...         ...
3969  ZUENOULA  In Service    INTERIEUR  ...   CENTRE NORTH  1cfc2d7ab1  f91ed8d2e9
3970  ZUENOULA  In Service    INTERIEUR  ...   CENTRE NORTH  3ae1708153  f91ed8d2e9
3971  ZUENOULA  In Service    INTERIEUR  ...   CENTRE NORTH  cb71f8a29c  61bee9c70c
3972  ZUENOULA  In Service    INTERIEUR  ...   CENTRE NORTH  d4a6a0b645  61bee9c70c
3973  ZUENOULA  In Service    INTERIEUR  ...   CENTRE NORTH  f9f3e42589  61bee9c70c

[3880 rows x 11 columns]

This step we now join the three datasets to the cell geo dataset.

In [30]:
#first we rename the columns for consistency

dfd1v1.columns = ['PRODUT', 'VALUE', 'DATETIME', 'CELL_ON_SITE', 'DW_A_NUMBER_INT',
       'DW_B_NUMBER_INT', 'COUNTRY_A', 'COUNTRY_B', 'CELLID', 'SITE_ID']



#For dataset2
dfd2v1.columns = ['PRODUT', 'VALUE', 'DATETIME', 'CELL_ON_SITE', 'DW_A_NUMBER_INT',
       'DW_B_NUMBER_INT', 'COUNTRY_A', 'COUNTRY_B', 'CELLID', 'SITE_ID']



#For dataset2
dfd3v1.columns = ['PRODUT', 'VALUE', 'DATETIME', 'CELL_ON_SITE', 'DW_A_NUMBER_INT',
       'DW_B_NUMBER_INT', 'COUNTRY_A', 'COUNTRY_B', 'CELLID', 'SITE_ID']

dfd1v1

PRODUT  VALUE               DATETIME  ... COUNTRY_B      CELLID     SITE_ID
0     Voice      0  2012-05-06 23:04:37.0  ...     19e2e  /TJNe+Mmtu  /TJNe+Mmtu
1       sms      0  2012-05-06 23:05:25.0  ...     19e2e  /+cKZKIp41  /+cKZKIp41
2       sms     25  2012-05-06 23:05:41.0  ...     19e2e  /u0FSD+ahi  /u0FSD+ahi
3     Voice     56  2012-05-06 23:05:42.0  ...     19e2e  +cNeJzsTp3  +cNeJzsTp3
4     Voice     14  2012-05-06 23:06:29.0  ...     19e2e  051KhYDCpv  051KhYDCpv
...     ...    ...                    ...  ...       ...         ...         ...
4996    sms    100  2012-05-07 00:01:54.0  ...     OTHER  0c2371d9bc  1e751d08a4
4997  Voice      1  2012-05-07 00:01:54.0  ...     19e2e  5fb791cf56  65efe2bc62
4998  Voice      0  2012-05-07 00:01:54.0  ...     19e2e  7773627c12  9ef9616350
4999  Voice      0  2012-05-07 00:01:54.0  ...     19e2e  3f4af49d6c  969b8d0e39
5000  Voice      7  2012-05-07 00:01:54.0  ...     19e2e  6f4d9b77a5  ebfc58af70

[4285 rows x 10 columns]

In [34]:
#Merge starting with the first dataset
#finaldf =pd.merge(combined_dfv4,dfcellsgeo, on= 'CELL_ID')
day1df = dfd1v1.join (dfcellsgeov1)
day1 = day1df.dropna()

<bound method NDFrame.head of      PRODUT  VALUE  ...     CELL_ID   SITE_CODE
23     data      0  ...  f740406877  0f34c36ac5
24    Voice      0  ...  959e54c8f7  0f34c36ac5
25     data      0  ...  56b7e7000c  0f34c36ac5
26    Voice      0  ...  bc51c17a6f  0f34c36ac5
27    Voice     19  ...  e460975c9f  0f34c36ac5
...     ...    ...  ...         ...         ...
3969  Voice     20  ...  1cfc2d7ab1  f91ed8d2e9
3970  Voice     30  ...  3ae1708153  f91ed8d2e9
3971  Voice      0  ...  cb71f8a29c  61bee9c70c
3972  Voice     29  ...  d4a6a0b645  61bee9c70c
3973  Voice     40  ...  f9f3e42589  61bee9c70c

[3329 rows x 21 columns]>

In [37]:
day2df = dfd2v1.join (dfcellsgeov1)
day2 = day2df.dropna()
day2

PRODUT  VALUE  ...     CELL_ID   SITE_CODE
23      sms      0  ...  f740406877  0f34c36ac5
24      sms      0  ...  959e54c8f7  0f34c36ac5
25    Voice      0  ...  56b7e7000c  0f34c36ac5
26      sms      0  ...  bc51c17a6f  0f34c36ac5
27      sms      0  ...  e460975c9f  0f34c36ac5
...     ...    ...  ...         ...         ...
3968    sms      0  ...  dcd3a41af8  f91ed8d2e9
3969    sms      0  ...  1cfc2d7ab1  f91ed8d2e9
3970    sms      0  ...  3ae1708153  f91ed8d2e9
3972  Voice     34  ...  d4a6a0b645  61bee9c70c
3973  Voice      0  ...  f9f3e42589  61bee9c70c

[3381 rows x 21 columns]

In [38]:
day3df = dfd3v1.join (dfcellsgeov1)
day3 = day3df.dropna()
day3

PRODUT  VALUE  ...     CELL_ID   SITE_CODE
23      sms     25  ...  f740406877  0f34c36ac5
24      sms      0  ...  959e54c8f7  0f34c36ac5
25    Voice      0  ...  56b7e7000c  0f34c36ac5
26      sms     25  ...  bc51c17a6f  0f34c36ac5
27     data      1  ...  e460975c9f  0f34c36ac5
...     ...    ...  ...         ...         ...
3969    sms      0  ...  1cfc2d7ab1  f91ed8d2e9
3970    sms      0  ...  3ae1708153  f91ed8d2e9
3971    sms     25  ...  cb71f8a29c  61bee9c70c
3972    sms     25  ...  d4a6a0b645  61bee9c70c
3973    sms      0  ...  f9f3e42589  61bee9c70c

[3412 rows x 21 columns]

In [40]:
#We now merge the three days to one dataset for analysis


combined_df = pd.concat([day1,day2,day3])
combined_df


PRODUT  VALUE  ...     CELL_ID   SITE_CODE
23     data      0  ...  f740406877  0f34c36ac5
24    Voice      0  ...  959e54c8f7  0f34c36ac5
25     data      0  ...  56b7e7000c  0f34c36ac5
26    Voice      0  ...  bc51c17a6f  0f34c36ac5
27    Voice     19  ...  e460975c9f  0f34c36ac5
...     ...    ...  ...         ...         ...
3969    sms      0  ...  1cfc2d7ab1  f91ed8d2e9
3970    sms      0  ...  3ae1708153  f91ed8d2e9
3971    sms     25  ...  cb71f8a29c  61bee9c70c
3972    sms     25  ...  d4a6a0b645  61bee9c70c
3973    sms      0  ...  f9f3e42589  61bee9c70c

[10122 rows x 21 columns]

#Analysis

In [41]:
#To analyze first we remove the unwanted columns

combined_df = combined_df.drop(columns = ['COUNTRY_A' ,	'COUNTRY_B' ,'CELL_ON_SITE'])
combined_df

PRODUT  VALUE  ...     CELL_ID   SITE_CODE
23     data      0  ...  f740406877  0f34c36ac5
24    Voice      0  ...  959e54c8f7  0f34c36ac5
25     data      0  ...  56b7e7000c  0f34c36ac5
26    Voice      0  ...  bc51c17a6f  0f34c36ac5
27    Voice     19  ...  e460975c9f  0f34c36ac5
...     ...    ...  ...         ...         ...
3969    sms      0  ...  1cfc2d7ab1  f91ed8d2e9
3970    sms      0  ...  3ae1708153  f91ed8d2e9
3971    sms     25  ...  cb71f8a29c  61bee9c70c
3972    sms     25  ...  d4a6a0b645  61bee9c70c
3973    sms      0  ...  f9f3e42589  61bee9c70c

[10122 rows x 18 columns]

#Which ones were the most used city for the three days?

In [42]:
result = pd.value_counts(combined_df['VILLES'])
result.head()


COCODY         942
YOPOUGON       746
ABOBO          694
MARCORY        371
TREICHVILLE    340
Name: VILLES, dtype: int64

COCODY is the most used city for the three days with 942 users for the tree days

#Which cities were the most used during business and home hours?

In [48]:
#First we need to convert dates to times we can manipulate

combined_df[['DATE', 'TIME']] = combined_df.DATETIME.str.split(expand=True)
combined_df

PRODUT  VALUE               DATETIME  ...   SITE_CODE        DATE        TIME
23     data      0  2012-05-06 23:11:27.0  ...  0f34c36ac5  2012-05-06  23:11:27.0
24    Voice      0  2012-05-06 23:11:27.0  ...  0f34c36ac5  2012-05-06  23:11:27.0
25     data      0  2012-05-06 23:12:07.0  ...  0f34c36ac5  2012-05-06  23:12:07.0
26    Voice      0  2012-05-06 23:12:10.0  ...  0f34c36ac5  2012-05-06  23:12:10.0
27    Voice     19  2012-05-06 23:12:25.0  ...  0f34c36ac5  2012-05-06  23:12:25.0
...     ...    ...                    ...  ...         ...         ...         ...
3969    sms      0  2012-05-09 00:01:16.0  ...  f91ed8d2e9  2012-05-09  00:01:16.0
3970    sms      0  2012-05-09 00:01:16.0  ...  f91ed8d2e9  2012-05-09  00:01:16.0
3971    sms     25  2012-05-09 00:01:16.0  ...  61bee9c70c  2012-05-09  00:01:16.0
3972    sms     25  2012-05-09 00:01:16.0  ...  61bee9c70c  2012-05-09  00:01:16.0
3973    sms      0  2012-05-09 00:01:16.0  ...  61bee9c70c  2012-05-09  00:01:16.0

[10122 rows x 20 columns]

In [68]:
busycities = combined_df.groupby(['PRODUT', 'VILLES', 'DATE', 'TIME']) ['VALUE'].sum()
busycities

PRODUT  VILLES       DATE        TIME      
Voice   ABENGOUROU   2012-05-06  23:15:34.0     14
                                 23:15:41.0    152
                                 23:15:44.0    138
                                 23:15:46.0     30
                                 23:16:02.0    192
                                              ... 
sms     ZON-HOUNIEN  2012-05-08  00:01:24.0      0
        ZOUKOUGBEU   2012-05-08  00:01:24.0     75
        ZUENOULA     2012-05-08  00:01:24.0      0
                     2012-05-09  00:01:01.0      0
                                 00:01:16.0     75
Name: VALUE, Length: 4013, dtype: int64

In [69]:
busycities = combined_df.groupby(['PRODUT', 'VILLES', 'DATE', 'TIME']) ['VALUE'].sum()
busycities.tail()

PRODUT  VILLES       DATE        TIME      
sms     ZON-HOUNIEN  2012-05-08  00:01:24.0     0
        ZOUKOUGBEU   2012-05-08  00:01:24.0    75
        ZUENOULA     2012-05-08  00:01:24.0     0
                     2012-05-09  00:01:01.0     0
                                 00:01:16.0    75
Name: VALUE, dtype: int64

From this we can say SMS is the least popular service compared to Calls.
And the most active home hours are before midnight.
